module/zfs/dsl_crypt.c

```diff
@@ -469,8 +469,10 @@ dsl_crypto_can_set_keylocation(const char *dsname, const char *keylocation)
 		goto out;
 
 	ret = dsl_dir_hold(dp, dsname, FTAG, &dd, NULL);
-	if (ret != 0)
+	if (ret != 0) {
+		dd = NULL;
 		goto out;
+	}
```

Comment:
My feeling is this whole thing could be significantly simplified by not feeling the need to rely on a `goto` for every exit path.  For example, something like the following.  And the same approach can be applied to `spa_keystore_load_wkey()`.

```c
int
dsl_crypto_can_set_keylocation(const char *dsname, const char *keylocation)
{
        dsl_dir_t *dd;
        dsl_pool_t *dp;
        int ret = 0;

        ret = dsl_pool_hold(dsname, FTAG, &dp);
        if (ret != 0)
                return (ret);

        ret = dsl_dir_hold(dp, dsname, FTAG, &dd, NULL);
        if (ret != 0) {
                dsl_pool_rele(dp, FTAG);
                return (ret);
        }

        /* if dd is not encrypted, the value may only be "none" */
        if (dd->dd_crypto_obj == 0) {
                if (strcmp(keylocation, "none") != 0)
                        ret = SET_ERROR(EACCES);

                goto out;
        }

        /* check for a valid keylocation for encrypted datasets */
        if (!zfs_prop_valid_keylocation(keylocation, B_TRUE)) {
                ret = SET_ERROR(EINVAL);
                goto out;
        }

        /* check that this is an encryption root */ 
        uint64_t rddobj;
        ret = dsl_dir_get_encryption_root_ddobj(dd, &rddobj);
        if (ret == 0 && rddobj != dd->dd_object)
                ret = SET_ERROR(EACCES);

out:
        dsl_dir_rele(dd, FTAG);
        dsl_pool_rele(dp, FTAG);

        return (ret);
}
```